File: java-ver/src/main/java/com/prvz/url_checker/web/ApiController.java
package com.prvz.url_checker.web;

import com.prvz.url_checker.model.CheckTask;
import com.prvz.url_checker.model.CheckTaskHistory;
import com.prvz.url_checker.service.TaskService;
import com.prvz.url_checker.web.dto.CheckTaskDto;
import com.prvz.url_checker.web.dto.CheckTaskHistoryDto;
import com.prvz.url_checker.web.dto.CreateTaskDto;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.util.UriComponentsBuilder;

import javax.validation.Valid;
import java.time.OffsetDateTime;

import static org.springframework.http.ResponseEntity.ok;

@RestController
@RequestMapping("/api")
@RequiredArgsConstructor
public class ApiController {

    private final TaskService taskService;

    @PostMapping
    public ResponseEntity<?> addTask(@Valid @RequestBody CreateTaskDto taskDto,
                                     UriComponentsBuilder uriComponentsBuilder) {
        CheckTask task = taskService.newTask(new CheckTask(
                OffsetDateTime.now(),
                null,
                taskDto.getInterval(),
                taskDto.getUrls()
        ));
        return ResponseEntity
                .created(uriComponentsBuilder
                        .path("/api/{id}")
                        .buildAndExpand(task.getId())
                        .toUri())
                .build();
    }

    @PostMapping("/stop")
    public void stopLastTask() {
        taskService.stopLast();
    }

    @GetMapping("/{id}")
    public ResponseEntity<CheckTaskDto> getTask(@PathVariable("id") String id) {
        return ok(toTaskDto(taskService.getTaskInfo(id)));
    }

    @GetMapping("/{id}/history")
    public Page<CheckTaskHistoryDto> getTaskHistory(@PathVariable("id") String id, Pageable pageable) {
        return taskService
                .getTaskHistory(id, pageable)
                .map(this::toTaskHistoryDto);
    }

    @GetMapping
    public ResponseEntity<Page<CheckTaskDto>> getAll(Pageable pageable) {
        return ok(taskService.getAll(pageable).map(this::toTaskDto));
    }

    @GetMapping("/last")
    public ResponseEntity<CheckTaskDto> getLast() {
        return ok(toTaskDto(taskService.getLastInfo()));
    }

    @GetMapping("/history")
    public Page<CheckTaskHistoryDto> getLastHistory(Pageable pageable) {
        return taskService.getLastHistory(pageable).map(this::toTaskHistoryDto);
    }

    private CheckTaskDto toTaskDto(CheckTask task) {
        return new CheckTaskDto(
                task.getId(),
                task.getCreateDate(),
                task.getCloseDate(),
                task.getInterval(),
                task.getUrls()
        );
    }

    private CheckTaskHistoryDto toTaskHistoryDto(CheckTaskHistory history) {
        return new CheckTaskHistoryDto(
                history.getId(),
                history.getTaskId(),
                history.getUrl(),
                history.getCheckDate(),
                history.isResult()
        );
    }

}

File: src/main/kotlin/com/prvz/url_checker/controller/ApiController.kt
package com.prvz.url_checker.controller

import com.prvz.url_checker.*
import com.prvz.url_checker.service.TaskService
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.http.ResponseEntity
import org.springframework.http.ResponseEntity.notFound
import org.springframework.http.ResponseEntity.ok
import org.springframework.web.bind.annotation.*
import org.springframework.web.util.UriComponentsBuilder
import java.time.OffsetDateTime
import javax.validation.Valid

@RestController
@RequestMapping("/api")
open class ApiController(
        private val taskService: TaskService
) {

    @PostMapping
    fun addTask(@Valid @RequestBody taskDto: CreateTaskDto,
                uriComponentsBuilder: UriComponentsBuilder): ResponseEntity<Any> {
        val task = taskService.newTask(CheckTask(
                createDate = OffsetDateTime.now(),
                interval = taskDto.interval,
                urls = taskDto.urls,
                checkType = CheckType.valueOf(taskDto.checkType.name)
        ))
        return ResponseEntity
                .created(uriComponentsBuilder
                        .path("/api/{id}")
                        .buildAndExpand(task.id)
                        .toUri())
                .build()

    }

    @PutMapping
    fun stopTask(@Valid @RequestBody taskDto: StopTaskDto?) {
        if (taskDto == null) {
            taskService.stopLast()
        } else {
            taskService.stopTask(taskDto.id)
        }
    }

    @GetMapping("/{id}")
    fun getTask(@PathVariable("id") id: String): ResponseEntity<CheckTaskDto> =
            taskService.getTaskInfo(id)
                    ?.let {
                        ok().body(CheckTaskDto(
                                id = it.id!!,
                                createDate = it.createDate,
                                closeDate = it.closeDate,
                                interval = it.interval,
                                urls = it.urls))
                    }
                    ?: notFound().build()

    @GetMapping("/{id}/history")
    fun getTaskHistory(@PathVariable("id") id: String, pageable: Pageable): Page<CheckTaskHistoryDto> =
            taskService.getTaskHistory(id, pageable)
                    .map { entity ->
                        CheckTaskHistoryDto(
                                id = entity.id!!,
                                taskId = entity.taskId,
                                url = entity.url,
                                checkDate = entity.checkDate,
                                result = entity.result
                        )
                    }


    @GetMapping
    fun test() {
        val task = CheckTask(
                createDate = OffsetDateTime.now(),
                interval = 10,
                urls = listOf("http://ya.ru", "https://mail.ru"),
                checkType = CheckType.BY_STATUS)
        taskService.newTask(task)
    }
}

